src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -60,58 +65,340 @@ public class ResponseCachingMiddleware
+            }
+        }
+
+        internal static async Task<bool> TryServeCachedResponseAsync(ResponseCachingContext context, CachedResponse cachedResponse)
+        {
+            context.CachedResponse = cachedResponse;
+            context.CachedResponseHeaders = new ResponseHeaders(cachedResponse.Headers);
+            context.ResponseTime = context.Options.SystemClock.UtcNow;
+            var cachedEntryAge = context.ResponseTime - context.CachedResponse.Created;
+            context.CachedEntryAge = cachedEntryAge > TimeSpan.Zero ? cachedEntryAge : TimeSpan.Zero;
 
-                // Should we attempt any caching logic?
-                if (_cacheabilityValidator.RequestIsCacheable(context))
+            if (context.CacheabilityValidator.IsCachedEntryFresh(context))
+            {
+                // Check conditional request rules
+                if (ConditionalRequestSatisfied(context))
+                {
+                    context.HttpContext.Response.StatusCode = StatusCodes.Status304NotModified;
+                }
+                else
                 {
-                    // Can this request be served from cache?
-                    if (await cachingContext.TryServeFromCacheAsync())
+                    var response = context.HttpContext.Response;
+                    // Copy the cached status code and response headers
+                    response.StatusCode = context.CachedResponse.StatusCode;
+                    foreach (var header in context.CachedResponse.Headers)
                     {
-                        return;
+                        response.Headers.Add(header);
                     }
 
-                    // Hook up to listen to the response stream
-                    cachingContext.ShimResponseStream();
+                    response.Headers[HeaderNames.Age] = context.CachedEntryAge.TotalSeconds.ToString("F0", CultureInfo.InvariantCulture);
+
+                    var body = context.CachedResponse.Body ??
+                        ((CachedResponseBody)context.Cache.Get(context.CachedResponse.BodyKeyPrefix))?.Body;
+
+                    // If the body is not found, something went wrong.
+                    if (body == null)
+                    {
+                        return false;
+                    }
 
-                    try
+                    // Copy the cached response body
+                    if (body.Length > 0)
                     {
-                        // Subscribe to OnStarting event
-                        context.Response.OnStarting(OnStartingCallback, cachingContext);
+                        // Add a content-length if required
+                        if (response.ContentLength == null && StringValues.IsNullOrEmpty(response.Headers[HeaderNames.TransferEncoding]))
+                        {
+                            response.ContentLength = body.Length;
+                        }
+                        await response.Body.WriteAsync(body, 0, body.Length);
```

Comment:
You could remove the `async` keyword on this method and just return the result here instead.